tsl/src/compression/compression_dml.c

```diff
@@ -648,6 +665,136 @@ batch_matches(RowDecompressor *decompressor, ScanKeyData *scankeys, int num_scan
+	if (validity)
+	{
+		const size_t n_vector_result_words = (arrow->length + 63) / 64;
+		for (size_t i = 0; i < n_vector_result_words; i++)
+		{
+			result[i] &= validity[i];
+		}
+	}
+	else
+	{
+		Assert(arrow->null_count == 0);
+	}
+}
+
+/* Look for default value match by checking the first result.
+ * Default value arrow arrays contain a single member so that the only result that matters.
+ * If we fail this check, it means the whole batch passed so we can bail immediately.
+ */
+static inline bool
+check_default_value_match(const uint64 *result)
+{
+	return result[0] & 1;
+}
+
+static bool
+batch_matches_vectorized(RowDecompressor *decompressor, ScanKeyData *scankeys, int num_scankeys,
+						 tuple_filtering_constraints *constraints, bool *skip_current_tuple)
+{
+	const int n_rows =
+		DatumGetInt32(decompressor->compressed_datums[decompressor->count_compressed_attindex]);
+	const int bitmap_bytes = sizeof(uint64) * ((n_rows + 63) / 64);
+	uint64 *restrict result =
+		MemoryContextAlloc(decompressor->per_compressed_row_ctx, bitmap_bytes);
+	uint64 dict_result[(GLOBAL_MAX_ROWS_PER_COMPRESSION + 63) / 64];
+	memset(result, 0xFF, bitmap_bytes);
+	bool default_value = false;
+	bool batch_failed = false;
+
+	for (int sk = 0; sk < num_scankeys; sk++)
+	{
+		ArrowArray *arrow =
+			decompress_single_column(decompressor, scankeys[sk].sk_attno, &default_value);
+
+		/* Handle null check */
+		if (scankeys[sk].sk_flags & SK_ISNULL)
+		{
+			vector_nulltest(arrow, IS_NULL, result);
+			if (default_value && !check_default_value_match(result))
+			{
+				batch_failed = true;
+				break;
+			}
+			continue;
+		}
+
+		VectorPredicate *predicate = get_vector_const_predicate(scankeys[sk].sk_func.fn_oid);
+
+		/* Handle non-dictionary compressed data */
+		if (!arrow->dictionary)
+		{
+			predicate(arrow, scankeys[sk].sk_argument, result);
+		}
+		else
+		{
+			/* Handle dictionary compressed data by decompressing the dictionary
+			 * first and then translating the results to actual results */
+			const size_t dict_rows = arrow->dictionary->length;
+			const size_t dict_result_words = (dict_rows + 63) / 64;
+			memset(dict_result, 0xFF, dict_result_words * 8);
+			predicate(arrow->dictionary, scankeys[sk].sk_argument, dict_result);
+			translate_bitmap_from_dictionary(arrow, dict_result, result);
+		}
+
+		apply_validity_bitmap(arrow, result);
+
+		if (default_value && !check_default_value_match(result))
+		{
+			batch_failed = true;
+			break;
+		}
+	}
+
+	if (batch_failed)
+	{
+		return false;
+	}
+
+	bool any_rows_pass = false;
+	bool all_rows_pass = true;
+	for (int i = 0; i < n_rows / 64; i++)
+	{
+		any_rows_pass = any_rows_pass || (result[i] != 0);
+		all_rows_pass = all_rows_pass && (~result[i] == 0);
+	}
+
+	if (n_rows % 64 != 0)
+	{
+		const uint64 last_word_mask = ~0ULL >> (64 - n_rows % 64);
+		any_rows_pass |= (result[n_rows / 64] & last_word_mask) != 0;
+		all_rows_pass &= ((~result[n_rows / 64]) & last_word_mask) == 0;
+	}
```

Comment:
This is just `get_vector_qual_summary`, right? Maybe using it directly would be more convenient.